toolchain/parse/state.def

```diff
@@ -227,63 +227,54 @@ CARBON_PARSE_STATE(CodeBlockFinish)
 
 // Handles a declaration name and parameters, such as `Foo[...](...)`.
 //
-// Allowed parameters:
-// - None: `Foo` only.
-// - Optional: `Foo`, `Foo(...)`, or `Foo[...](...)`.
-// - Required: `Foo(...)` or `Foo[...](...)`.
-//
 // name . ...
-// ^~~~
-//   1. PeriodAsDecl
-//   2. DeclNameAndParamsAfterNameAs(None|Optional|Required)
+// ^~~~~~
+//   1. DeclNameAndParams
 //
-// name ...
+// name [ ... ]
 // ^~~~
-//   1. DeclNameAndParamsAfterNameAs(None|Optional|Required)
-//
-//  ???
-// ^
-//   (state done)
-CARBON_PARSE_STATE_VARIANTS3(DeclNameAndParams, None, Optional, Required)
-
-// Handles a declaration name between the main name and implicit parameters.
-//
-// name . ...
-//     ^
-//   1. PeriodAsDecl
-//   2. DeclNameAndParamsAfterNameAs(None|Optional|Required)
-//
-// name [ ... ]    (variant is not None)
-//     ^
 //   1. PatternListAsImplicit
 //   2. DeclNameAndParamsAfterImplicit
 //
-// name ( ... )    (variant is not None)
-//     ^
+// name ( ... )
+// ^~~~
 //   1. PatternListAsTuple
+//   2. DeclNameAndParamsAfterParams
 //
-// name ...        (variant is not Required)
-//     ^
+// name ...
+// ^~~~
 //   (state done)
 //
-// name ???        (variant is Required)
-//     ^
+//  ???
+// ^
 //   (state done)
-CARBON_PARSE_STATE_VARIANTS3(DeclNameAndParamsAfterName, None, Optional,
-                             Required)
+CARBON_PARSE_STATE(DeclNameAndParams)
 
 // Handles regular parameters such as `(...)` for the general declaration case.
 // Only used after implicit parameters.
 //
 // name [ ... ] ( ... )
 //             ^
 //   1. PatternListAsTuple
+//   1. DeclNameAndParamsAfterParams
 //
 // name [ ... ] ???
 //             ^
 //   (state done)
 CARBON_PARSE_STATE(DeclNameAndParamsAfterImplicit)
 
+// Handles regular parameters such as `(...)` for the general declaration case.
+// Only used after implicit parameters.
+//
+// name [ ... ] ( ... ) .
+//                      ^
+//   1. DeclNameAndParams
+//
+// name [ ... ] ( ... ) ...
+//                     ^
+//   (state done)
```

Comment:
I'm confused, how are these two inputs different? The `.` versus `...` looks like a typo.

Reply:
The first rule says: if the next token is a `.`, then consume it (because it's underlined), and continue in state `DeclNameAndParams`.

The second rule says: if the next token is anything else (per the comment at the start of the file: "`...` indicates a sequence of tokens handled by some other state"), then consume nothing and we're done.

Reply:
Okay, but maybe the comment could be clearer? Maybe this is a problem with using `...` without any extra annotation, it makes `.` look like a typo.